cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -216,6 +227,75 @@ import           Cardano.Ledger.Crypto (StandardCrypto)
 {- HLINT ignore "Redundant flip" -}
 {- HLINT ignore "Use section" -}
 
+-- | Indicates whether a script is expected to fail or pass validation.
+data ScriptValidity
+  = ScriptInvalid -- ^ Script is expected to fail validation.
+                  -- Transactions marked as such can include scripts that fail validation.
+                  -- Such transactions may be submitted to the chain, in which case the
+                  -- collateral will be taken upon on chain script validation failure.
+
+  | ScriptValid   -- ^ Script is expected to pass validation.
+                  -- Transactions marked as such cannot include scripts that fail validation.
+
+  deriving (Eq, Show)
+
+instance ToCBOR ScriptValidity where
+  toCBOR = toCBOR . scriptValidityToIsValid
+
+instance FromCBOR ScriptValidity where
+  fromCBOR = isValidToScriptValidity <$> fromCBOR
+
+scriptValidityToIsValid :: ScriptValidity -> Alonzo.IsValid
+scriptValidityToIsValid ScriptInvalid = Alonzo.IsValid False
+scriptValidityToIsValid ScriptValid = Alonzo.IsValid True
+
+isValidToScriptValidity :: Alonzo.IsValid -> ScriptValidity
+isValidToScriptValidity (Alonzo.IsValid False) = ScriptInvalid
+isValidToScriptValidity (Alonzo.IsValid True) = ScriptValid
+
+-- | A representation of whether the era supports tx script validity.
+--
+-- The Mary and subsequent eras support script validity.
+--
+data TxBodyScriptValidity era where
+  TxBodyScriptValidityNone :: TxBodyScriptValidity era
+
+  -- | Tx script validity is supported in transactions in the 'Alonzo' era onwards.
+  TxBodyScriptValidity
+    :: TxBodyScriptValiditySupportedInEra era
+    -> ScriptValidity
+    -> TxBodyScriptValidity era
+
+deriving instance Eq   (TxBodyScriptValiditySupportedInEra era)
+deriving instance Show (TxBodyScriptValiditySupportedInEra era)
+
+data TxBodyScriptValiditySupportedInEra era where
+  TxBodyScriptValiditySupportedInAlonzoEra :: TxBodyScriptValiditySupportedInEra AlonzoEra
+
+deriving instance Eq   (TxBodyScriptValidity era)
+deriving instance Show (TxBodyScriptValidity era)
+
+txBodyScriptValiditySupportedInEra :: CardanoEra era -> Maybe (TxBodyScriptValiditySupportedInEra era)
+txBodyScriptValiditySupportedInEra ByronEra   = Nothing
+txBodyScriptValiditySupportedInEra ShelleyEra = Nothing
+txBodyScriptValiditySupportedInEra AllegraEra = Nothing
+txBodyScriptValiditySupportedInEra MaryEra    = Nothing
+txBodyScriptValiditySupportedInEra AlonzoEra  = Just TxBodyScriptValiditySupportedInAlonzoEra
```

Comment:
Can we optimise this to:

```
txBodyScriptValiditySupportedInEra AlonzoEra  = Just TxBodyScriptValiditySupportedInAlonzoEra
txBodyScriptValiditySupportedInEra _ = Nothing
```
?

IIRC GHC compiles case matching alternative by alternative, in user-supplied order, so having a series of `Nothing` match failures would be quite inefficient?

Reply:
I'm wary of doing that because later if we add additional eras, the compiler won't catch that this function needs to change.

Reply:
>the compiler won't catch that this function needs to change.
This is more important because a lot of functions need to be updated when we add a new era and GHC tells us.